updates per writing style (#70978)
ansible/ansible: docs/docsite/rst/dev_guide/developing_modules_general.rst

.. _developing_modules_general:
.. _module_dev_tutorial_sample:

**************************
Developing Ansible modules
**************************

A module is a reusable, standalone script that Ansible runs on your behalf, either locally or remotely. Modules interact with your local machine, an API, or a remote system to perform specific tasks like changing a database password or spinning up a cloud instance. Each module can be used by the Ansible API, or by the :command:`ansible` or :command:`ansible-playbook` programs. A module provides a defined interface, accepts arguments, and returns information to Ansible by printing a JSON string to stdout before exiting. 

> Ansible ships thousands of default modules, and you can easily write your own custom module. If you're writing a module for local use, you can choose any programming language and follow your own rules. Use this topic to know how to create an Ansible module in Python. After you create a module, you must add it locally to the appropriate directory so that Ansible can find and execute it. For details about adding a module locally, see :ref:`developing_locally`.

Reviewer: I'd rephrase this - partly because of the collections migration and partly because the phrase "default modules" isn't fully accurate in the Ansible context . . .
```suggestion
If you need functionality that is not available in any of the thousands of Ansible modules found in collections, you can easily write your own custom module. When you write a module for local use, you can choose any programming language and follow your own rules. Use this topic to learn how to create an Ansible module in Python. After you create a module, you must add it locally to the appropriate directory so that Ansible can find and execute it. For details about adding a module locally, see :ref:`developing_locally`.
```

.. _module_dev_tutorial_sample:

*******************************************

> Developing Ansible modules

Reviewer 2: ```suggestion
**************************
Developing Ansible modules
**************************
```

*******************************************

A module is a reusable, standalone script that Ansible runs on your behalf, either locally or remotely. Modules interact with your local machine, an API, or a remote system to perform specific tasks like changing a database password or spinning up a cloud instance. Each module can be used by the Ansible API, or by the :command:`ansible` or :command:`ansible-playbook` programs. A module provides a defined interface, accepts arguments, and returns information to Ansible by printing a JSON string to stdout before exiting. 

Ansible ships thousands of default modules, and you can easily write your own custom module. If you're writing a module for local use, you can choose any programming language and follow your own rules. Use this topic to know how to create an Ansible module in Python. After you create a module, you must add it locally to the appropriate directory so that Ansible can find and execute it. For details about adding a module locally, see :ref:`developing_locally`.

.. contents:: Topics
   :local:

.. _environment_setup:

> Preparing an environment for developing Ansible modules

Reviewer 2: ```suggestion
Preparing an environment for developing Ansible modules
=======================================================
```

=================

> Installing prerequisites via apt (Ubuntu)

Reviewer 2: ```suggestion
Installing prerequisites via apt (Ubuntu)
-----------------------------------------
```

.. _module_dev_tutorial_sample:

> *******************************************

Reviewer: I think all the current errors are caused by the two sets of heading markers here. Somehow instead of replacing the longer markers, you added a new pair of them . . .
```suggestion
```

**************************
Developing Ansible modules
**************************

> *******************************************

Reviewer: This one also needs to be deleted:
```suggestion
```

    sudo apt update
    sudo apt install build-essential libssl-dev libffi-dev python-dev

> Creating a common development environment

Reviewer 2: ```suggestion
Creating a common development environment
-----------------------------------------
```

    sudo apt update
    sudo apt install build-essential libssl-dev libffi-dev python-dev

> Creating a common development environment

Reviewer: The original heading meant "Environment set-up steps that are common across most/all platforms" (as opposed to steps that are specific to Ubuntu). The revised heading changes the meaning. Maybe this could be "Creating a development environment (platform-agnostic steps)"?

   ``$ . venv/bin/activate && . hacking/env-setup``

> Creating a module

Reviewer 2: ```suggestion
Creating a module
=================
```

        main()

> Verifying your module code

Reviewer 2: ```suggestion
Verifying your module code
==========================
```

===========================

Once you've modified the sample code above to do what you want, you can try out your module.
Our :ref:`debugging tips <debugging>` will help if you run into bugs as you verify your module code.

> Verifying your module code locally

Reviewer 2: ```suggestion
Verifying your module code locally
----------------------------------
```

    {"changed": true, "state": {"original_message": "hello", "new_message": "goodbye"}, "invocation": {"module_args": {"name": "hello", "new": true}}}

> Verifying your module code in a playbook

Reviewer 2: ```suggestion
Verifying your module code in a playbook
----------------------------------------
```

- Run the playbook and analyze the output: ``$ ansible-playbook ./testmod.yml``

Testing your newly-created module

> ====================

Reviewer 3: expand header to match size

- Run the playbook and analyze the output: ``$ ansible-playbook ./testmod.yml``

Testing your newly-created module

> ====================

Reviewer: ```suggestion
=================================
```

The following two examples will get you started with testing your module code. Please review our :ref:`testing <developing_testing>` section for more detailed
information, including instructions for :ref:`testing module documentation <testing_module_documentation>`, adding :ref:`integration tests <testing_integration>`, and more.

Performing sanity tests

> ------------

Reviewer: ```suggestion
-----------------------
```

You can run through Ansible's sanity checks in a container:

``$ ansible-test sanity -v --docker --python 2.7 MODULE_NAME``

.. note:: 
	Note that this example requires Docker to be installed and running. If you'd rather not use a container for this, you can choose to use ``--venv`` instead of ``--docker``.

Performing unit tests

> ----------

Reviewer: ```suggestion
---------------------
```

> Performing unit tests

Reviewer: The sanity tests already exist, but for a new module, the developer must add new unit tests.
```suggestion
Adding unit tests
```
